source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,613 @@
+        context.logger.error("Releasing pointer({}) is not match to {}", Ptr,
+                             AllocInfo->UserBegin);
+        return UR_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    // TODO: Update shadow memory
+    return context.urDdiTable.USM.pfnFree(Context,
+                                          (void *)AllocInfo->AllocBegin);
+}
+
+bool SanitizerInterceptor::preLaunchKernel(ur_kernel_handle_t Kernel,
+                                           ur_queue_handle_t Queue,
+                                           ur_event_handle_t &Event) {
+    prepareLaunch(Queue, Kernel);
+
+    UR_CALL(updateShadowMemory(Queue));
+
+    // Return LastEvent in QueueInfo
+    auto Context = getContext(Queue);
+    auto ContextInfo = getContextInfo(Context);
+    auto QueueInfo = ContextInfo->getQueueInfo(Queue);
+
+    std::scoped_lock<ur_mutex> Guard(QueueInfo->Mutex);
+    Event = QueueInfo->LastEvent;
+    QueueInfo->LastEvent = nullptr;
+
+    return true;
+}
+
+void SanitizerInterceptor::postLaunchKernel(ur_kernel_handle_t Kernel,
+                                            ur_queue_handle_t Queue,
+                                            ur_event_handle_t *Event) {
+    auto Program = getProgram(Kernel);
+    ur_event_handle_t ReadEvent{};
+
+    // If kernel has defined SPIR_DeviceSanitizerReportMem, then we try to read it
+    // to host, but it's okay that it isn't defined
+    auto Result = context.urDdiTable.Enqueue.pfnDeviceGlobalVariableRead(
+        Queue, Program, kSPIR_DeviceSanitizerReportMem, true,
+        sizeof(SPIR_DeviceSanitizerReportMem), 0,
+        &SPIR_DeviceSanitizerReportMem, 1, Event, &ReadEvent);
+
+    if (Result == UR_RESULT_SUCCESS) {
+        *Event = ReadEvent;
+
+        auto AH = &SPIR_DeviceSanitizerReportMem;
+        if (!AH->Flag) {
+            return;
+        }
+
+        const char *File = AH->File[0] ? AH->File : "<unknown file>";
+        const char *Func = AH->Func[0] ? AH->Func : "<unknown func>";
+
+        context.logger.always("====ERROR: DeviceSanitizer: %s on %s",
+                              DeviceSanitizerFormat(AH->ErrorType),
+                              DeviceSanitizerFormat(AH->MemoryType));
+        context.logger.always(
+            "%s of size %u at kernel <%s> LID(%lu, %lu, %lu) GID(%lu, "
+            "%lu, %lu)",
+            AH->IsWrite ? "WRITE" : "READ", AH->AccessSize, Func, AH->LID0,
+            AH->LID1, AH->LID2, AH->GID0, AH->GID1, AH->GID2);
+        context.logger.always("  #0 %s %s:%d", Func, File, AH->Line);
+        if (!AH->IsRecover) {
+            abort();
+        }
+    }
+}
+
+std::string SanitizerInterceptor::getKernelName(ur_kernel_handle_t Kernel) {
+    size_t KernelNameSize = 0;
+    [[maybe_unused]] auto Res = context.urDdiTable.Kernel.pfnGetInfo(
+        Kernel, UR_KERNEL_INFO_FUNCTION_NAME, 0, nullptr, &KernelNameSize);
+    assert(Res == UR_RESULT_SUCCESS);
+
+    std::vector<char> KernelNameBuf(KernelNameSize + 1);
```

Comment:
I'm 99% sure you can just do `KernelNameBuf(KernelNameSize)`, the spec for `UR_KERNEL_INFO_FUNCTION_NAME` states "Return null-terminated kernel function name.". Unless the drivers you're testing with aren't adding the null character?

Reply:
Makes sense. I tested it and you're right. Thanks!